### ISSUE TITLE
weights decay and constant biases

### ISSUE THREAD

**Issue by tlkvstepan** [ClementPinard/FlowNetPytorch#16]
Hello! It seems that authors use weight decay 4e-4. It is used only for weight and not for biases, and only for contracting part.  Have you tried it?

```bash
layer {
  name: "conv1"
  type: "Convolution"
  bottom: "input"
  top: "conv1"
  param {
    lr_mult: 1
    **decay_mult: 1**
  }
  param {
    lr_mult: 1
    **decay_mult: 0**
  }
  convolution_param {
    num_output: 64
    pad: 3
    kernel_size: 7
    stride: 2
    weight_filler {
      type: "msra"
    }
    bias_filler {
      type: "constant"
    }
    engine: CUDNN
  }
}
```

**Comment by tlkvstepan** [comment]
And one more thing, in contrast to your code, in the original code there is non biases in deconv blocks, but there are biases in conv blocks. Is it implementation error or network behaves better in this way?

> layer {
>   name: "deconv3"
>   type: "Deconvolution"
>   bottom: "concat4"
>   top: "deconv3"
>   param {
>     lr_mult: 1
>     decay_mult: 0
>   }
>   param {
>     **lr_mult: 0**
>     decay_mult: 0
>   }
>   convolution_param {
>     num_output: 128
>     pad: 1
>     kernel_size: 4
>     stride: 2
>     weight_filler {
>       type: "msra"
>     }
>     bias_filler {
>       type: "constant"
>     }
>     engine: CUDNN
>   }
> }

> layer {
>   name: "conv3"
>   type: "Convolution"
>   bottom: "conv2"
>   top: "conv3"
>   param {
>     lr_mult: 1
>     decay_mult: 1
>   }
>   param {
>     **lr_mult: 1**
>     decay_mult: 0
>   }
>   convolution_param {
>     num_output: 256
>     pad: 2
>     kernel_size: 5
>     stride: 2
>     weight_filler {
>       type: "msra"
>     }
>     bias_filler {
>       type: "constant"
>     }
>     engine: CUDNN
>   }
> }

**Comment by ClementPinard** [comment]
I believe the `bias_filler` is just for initialization : you begin with 0 biases and then let it update. 
Where do you see in these parameters that the bias is treated differently between conv and deconv ? I see the same bias filler for both.

As for weight decay, it could be a thing to test. I am not particularly familiar with layer params in caffe, are the two sequential params applied to weights and then bias ?

**Comment by tlkvstepan** [comment]
> are  the two sequential params applied to weights and then bias ?

Yes, I think so. See description in caffe [Caffe] (http://caffe.berkeleyvision.org/tutorial/layers/convolution.html)

> I believe the bias_filler is just for initialization : you begin with 0 biases and then let it update.
> Where do you see in these parameters that the bias is treated differently between conv and deconv ? I see the same bias filler for both.

Please look at the param field for deconv layer... It has lr_mult.. I assume this is learning rate multiplier. It is zero for biases. In other word they are not updated during training.

**Comment by ClementPinard** [comment]
You are indeed right, What's more, when I check the bias of pretrained for deconvolution the bias are all set to 0.
This is great news as I did not reach a sufficient EPE yet. I'm going to change the implementation accordingly.

As for weight decay, applying it only onweights is a bit hard for pytorch since it is set directly on the optimizer. I'm going to try it with the same for everything before and see where it goes.

Otherwise, I will have to designe a L2 loss for weights only and that will be ugly (but maybe useful, we'll see)

Thanks a lot for your help !

I am honestly surprised about the attention this project got lately as it is almost 1 year old, but lots of people are helping figuring out what exactly is done in caffe training to replicate it, that's great :smile:

**Comment by tlkvstepan** [comment]
``` As for weight decay, applying it only onweights is a bit hard for pytorch since it is set directly on the optimizer. I'm going to try it with the same for everything before and see where it goes. ```

You could use per parameter optimizer option (divide parameters into two groups).

I have one more question for you. When you train the net with augmentation, do you get same convergence speed? For some reason in my experiments network converges much faster (in terms of epochs number) if I don't use augmentation..  Does augmentation helps in your case?

**Comment by ClementPinard** [comment]
I have update the code, which is now dealing with bias less deconvolutions, and bias / weight segmentation for weight decay !

As for convergence speed, I did not notice anything particular regarding data augmentation. I plan to completely change it, as dealing with affine matrices instead of translation + rotation composition just like in original caffe code is much more clever for Flow warping.
